pyintegration/tests/test_dtypes.py

```diff
@@ -0,0 +1,181 @@
+            j_array = dtypes.byte.array_from(expected)
+            self.assertIn("[B", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertNotEqual(expected, py_array)
+
+    def test_floating_array_of(self):
+
+        nulls = {dtypes.float_: NULL_FLOAT, dtypes.double: NULL_DOUBLE}
+
+        np_array = np.array([float('nan'), 1.7976931348623157e+300, NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+        for dt, nv in nulls.items():
+            map_fn = functools.partial(remap_double, null_value=nv)
+            with self.subTest(f"numpy double array to {dt} with mapping"):
+                expected = [nv, 1.7976931348623157e+300, nv, 1.1, nv]
+                j_array = dt.array_from(np_array, remap=map_fn)
+                py_array = [x for x in j_array]
+                for i in range(4):
+                    # downcast from double to float results in inf when the value is outside of float range
+                    self.assertTrue(math.isclose(expected[i], py_array[i], rel_tol=1e-7) or py_array[i] == float('inf'))
+
+        with self.subTest("double array from numpy array"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            pd_series = pd.Series(np_array)
+            j_array = dtypes.double.array_from(pd_series)
+            py_array = [x for x in j_array]
+            expected = [float('nan'), NULL_DOUBLE, 1.1, float('inf')]
+            self.assertTrue(math.isnan(py_array[0]))
+            self.assertEqual(expected[1:], py_array[1:])
+
+        with self.subTest("double array from numpy long array"):
+            expected = [NULL_LONG, 1, 2, 3]
+            np_array = np.array(expected, dtype=np.int64)
+            j_array = dtypes.float64.array_from(np_array)
+            self.assertIn("[D", str(type(j_array)))
+            py_array = [x for x in j_array]
+            for i in range(4):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+        with self.subTest("double array from Python list of integer"):
+            expected = [NULL_LONG, 1, 2, 3]
+            j_array = dtypes.float64.array_from(expected)
+            py_array = [x for x in j_array]
+            for i in range(3):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+    def test_char_array_of(self):
+        def remap_char(v):
+            if v is None:
+                return NULL_CHAR
+            if isinstance(v, int):
+                return v
+            if isinstance(v, str):
+                if len(v) < 1:
+                    return NULL_CHAR
+                return ord(v[0])
+            try:
+                return int(v)
+            except:
+                return NULL_CHAR
+
+        test_str = "abcdefg0123456"
+        j_array = dtypes.char.array_from(test_str)
+        self.assertIn("[C", str(type(j_array)))
+        py_array = [chr(x) for x in j_array]
+        self.assertEqual(test_str, "".join(py_array))
```

Comment:
I don't love this assertion.  Ideally, you want to check that `[a,b,c,d]` equals `[a,b,c,d]`.  With the join in the assertion, you could have something like `[abcd,"","",""]` passing the assertion as equal to `[a,b,c,d,]`.

Reply:
sure, but the next seertEqual covers that.